File: ios/CoachCore/Sources/CoachCore/ExerciseSpec.swift
// Exercise definitions — thresholds identical to the desktop prototype.

import Foundation

public enum ExerciseMode { case reps, hold }

public struct ExerciseSpec {
    public let name: String
    public let signal: SignalKey       // angle driving the FSM (down, then up)
    public let startBelow: Double      // below this => rep started
    public let bottomBelow: Double     // deep enough for full ROM
    public let lockoutAbove: Double    // back above this => rep complete
    public let concentricPhase: String // "ascent" | "descent" (angle direction of the lift)
    public let minRepS: Double
    public let minConcentricS: Double  // faster => "slow down"
    public let mode: ExerciseMode
    public let cameraHint: String

    public init(name: String, signal: SignalKey,
                startBelow: Double = 0, bottomBelow: Double = 0,
                lockoutAbove: Double = 0, concentricPhase: String = "ascent",
                minRepS: Double = 0.8, minConcentricS: Double = 0.6,
                mode: ExerciseMode = .reps,
                cameraHint: String? = nil) {
        self.name = name
        self.signal = signal
        self.startBelow = startBelow
        self.bottomBelow = bottomBelow
        self.lockoutAbove = lockoutAbove
        self.concentricPhase = concentricPhase
        self.minRepS = minRepS
        self.minConcentricS = minConcentricS
        self.mode = mode
        self.cameraHint = cameraHint ?? loc("hint.side")
    }
}

public let specs: [String: ExerciseSpec] = [
    "squat": ExerciseSpec(name: "squat", signal: .knee, startBelow: 150,
                          bottomBelow: 100, lockoutAbove: 165,
                          cameraHint: loc("hint.side_45front")),
    "pushup": ExerciseSpec(name: "pushup", signal: .elbow, startBelow: 140,
                           bottomBelow: 95, lockoutAbove: 155,
                           minConcentricS: 0.4),
    "bench": ExerciseSpec(name: "bench", signal: .elbow, startBelow: 140,
                          bottomBelow: 90, lockoutAbove: 160,
                          cameraHint: loc("hint.side_head")),
    "deadlift": ExerciseSpec(name: "deadlift", signal: .hip, startBelow: 150,
                             bottomBelow: 100, lockoutAbove: 165),
    "lunge": ExerciseSpec(name: "lunge", signal: .knee, startBelow: 150,
                          bottomBelow: 110, lockoutAbove: 165,
                          cameraHint: loc("hint.side_45knee")),
    "shoulder_press": ExerciseSpec(name: "shoulder_press", signal: .elbow,
                                   startBelow: 150, bottomBelow: 100,
                                   lockoutAbove: 160,
                                   cameraHint: loc("hint.front")),
    "curl": ExerciseSpec(name: "curl", signal: .elbow, startBelow: 140,
                         bottomBelow: 70, lockoutAbove: 155,
                         concentricPhase: "descent", minConcentricS: 0.5),
    "pullup": ExerciseSpec(name: "pullup", signal: .elbow, startBelow: 140,
                           bottomBelow: 80, lockoutAbove: 160,
                           concentricPhase: "descent",
                           cameraHint: loc("hint.front")),
    "plank": ExerciseSpec(name: "plank", signal: .bodyLine, mode: .hold),
]

/// Display order for pickers.
public let exerciseOrder = ["squat", "pushup", "bench", "deadlift", "lunge",
                            "shoulder_press", "curl", "pullup", "plank"]

public func displayName(_ exercise: String) -> String {
    let key = "exercise.\(exercise)"
    let localized = NSLocalizedString(key, bundle: .module, comment: "")
    if localized != key { return localized }
    return exercise.replacingOccurrences(of: "_", with: " ").capitalized
}

File: android/app/src/main/java/com/fekitech/gymcoach/core/ExerciseSpec.kt
// Exercise definitions — thresholds identical to the desktop prototype
// and iOS CoachCore.
package com.fekitech.gymcoach.core

enum class ExerciseMode { REPS, HOLD }

data class ExerciseSpec(
    val name: String,
    val signal: SignalKey,            // angle driving the FSM (down, then up)
    val startBelow: Double = 0.0,     // below this => rep started
    val bottomBelow: Double = 0.0,    // deep enough for full ROM
    val lockoutAbove: Double = 0.0,   // back above this => rep complete
    val concentricPhase: String = "ascent", // angle direction of the lift
    val minRepS: Double = 0.8,
    val minConcentricS: Double = 0.6, // faster => "slow down"
    val mode: ExerciseMode = ExerciseMode.REPS,
    val cameraHint: String = "Place the phone to see your full body from the side.",
)

val specs: Map<String, ExerciseSpec> = mapOf(
    "squat" to ExerciseSpec("squat", SignalKey.KNEE, startBelow = 150.0,
        bottomBelow = 100.0, lockoutAbove = 165.0,
        cameraHint = "Side view or 45°, whole body in frame."),
    "pushup" to ExerciseSpec("pushup", SignalKey.ELBOW, startBelow = 140.0,
        bottomBelow = 95.0, lockoutAbove = 155.0, minConcentricS = 0.4),
    "bench" to ExerciseSpec("bench", SignalKey.ELBOW, startBelow = 140.0,
        bottomBelow = 90.0, lockoutAbove = 160.0,
        cameraHint = "Side view, head to hips in frame."),
    "deadlift" to ExerciseSpec("deadlift", SignalKey.HIP, startBelow = 150.0,
        bottomBelow = 100.0, lockoutAbove = 165.0),
    "lunge" to ExerciseSpec("lunge", SignalKey.KNEE, startBelow = 150.0,
        bottomBelow = 110.0, lockoutAbove = 165.0,
        cameraHint = "Side view or 45°, front knee visible."),
    "shoulder_press" to ExerciseSpec("shoulder_press", SignalKey.ELBOW,
        startBelow = 150.0, bottomBelow = 100.0, lockoutAbove = 160.0,
        cameraHint = "Face the camera, arms fully in frame."),
    "curl" to ExerciseSpec("curl", SignalKey.ELBOW, startBelow = 140.0,
        bottomBelow = 70.0, lockoutAbove = 155.0,
        concentricPhase = "descent", minConcentricS = 0.5),
    "pullup" to ExerciseSpec("pullup", SignalKey.ELBOW, startBelow = 140.0,
        bottomBelow = 80.0, lockoutAbove = 160.0,
        concentricPhase = "descent",
        cameraHint = "Face the camera, bar and body in frame."),
    "plank" to ExerciseSpec("plank", SignalKey.BODY_LINE,
        mode = ExerciseMode.HOLD),
)

/** Display order for pickers. */
val exerciseOrder = listOf("squat", "pushup", "bench", "deadlift", "lunge",
                           "shoulder_press", "curl", "pullup", "plank")

fun displayName(exercise: String): String =
    exercise.replace('_', ' ')
        .split(' ')
        .joinToString(" ") { w -> w.replaceFirstChar { it.uppercase() } }
